### DISCUSSION TITLE
Query related to page walk caches and iommu

### DISCUSSION THREAD

**Discussion by Kranti-Ingale** [sarchlab/mgpusim#122] [Q&A]
Hi,

I am a beginner currently learning to use MGPUSIM and exploring its integration with the Akita framework . I have a couple of queries and would be truly grateful for your guidance:.
1) Is the driver component of MGPUSIM essentially an implementation of the IOMMU?
2) I noticed that there doesn’t seem to be a page walk cache in MGPUSIM or Akita. Is this something I need to code myself, or am I possibly missing an existing implementation?

**Comment by syifan** [comment]
1. Driver does not implment IOMMU. The driver is the controller that determines what the GPU needs to do. The MMU (https://github.com/sarchlab/akita/blob/v3/mem/vm/mmu/mmu.go) implements IOMMU. 
2. We do not. Our MMU is a really simple (fixed latency) model. So, if you need to have a more accurate model, you need to implement with new code. But you can discuss what you want to implement, and we can provide you with some guidance.

**Reply by Kranti-Ingale** [reply]
Thank you for the clarifications. 

Each GMMU has a Page Walk Cache (PWC). However, when I checked the **gmmu** folder in *akita*, I found implementations for the Page Table but not for the PWC.

**Reply by syifan** [reply]
@DX990307 Could you help answer this question?

**Reply by DX990307** [reply]
Yes, in our implementation, there is no page walk cache in GMMU. What we did is add a TLB as a page walk cache between the GMMU and L2 TLB.

**Reply by Kranti-Ingale** [reply]
Thank you for the clarification. I just want to confirm my understanding based on your response:

In the distributed address translation process, traditionally, if there is an L1 TLB miss and an L2 TLB miss, the request proceeds to the GMMU. At this point, the GMMU typically checks the Page Walk Cache (PWC) before initiating the full page walk to the GPU Page Table (PT).

However, from your explanation, Does this mean that:

1. The TLB you added is different from the traditional L1 TLBs?
2. During a TLB miss (L1 and L2), the GMMU will first query this new TLB (acting as a page walk cache) before initiating the page walk to access the GPU's page table?

Thank you for your time and patience!

**Reply by DX990307** [reply]
I think in hardware, the page table cache has some differences from the traditional TLBs. For the details, you can check this paper [https://kib.kiev.ua/x86docs/Intel/WhitePapers/317080-002.pdf]. But in our simulator, we simplify the page table cache as the L3 TLB since they have the same function.

**Reply by Kranti-Ingale** [reply]
I can only see the L1 and L2 TLB implementation in TLB folder. In GMMU folder there is only Page Table implementation. can you please share exact location of folder where to refer L3TLB.

**Reply by syifan** [reply]
Support for an L3 TLB is still some ongoing work. But it should be done soon. You can check here. https://github.com/sarchlab/mgpusim/pull/124

**Reply by Kranti-Ingale** [reply]
Thank You sir and @DX990307
